// pgnode/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tx = exports.types = exports.native = exports.Query = exports.Pool = exports.Events = exports.Connection = exports.ClientBase = exports.Client = void 0;
const pg_1 = __importDefault(require("pg"));
const core_1 = __importDefault(require("./core"));
exports.tx = core_1.default;
var pg_2 = require("pg");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return pg_2.Client; } });
Object.defineProperty(exports, "ClientBase", { enumerable: true, get: function () { return pg_2.ClientBase; } });
Object.defineProperty(exports, "Connection", { enumerable: true, get: function () { return pg_2.Connection; } });
Object.defineProperty(exports, "Events", { enumerable: true, get: function () { return pg_2.Events; } });
Object.defineProperty(exports, "Pool", { enumerable: true, get: function () { return pg_2.Pool; } });
Object.defineProperty(exports, "Query", { enumerable: true, get: function () { return pg_2.Query; } });
Object.defineProperty(exports, "native", { enumerable: true, get: function () { return pg_2.native; } });
Object.defineProperty(exports, "types", { enumerable: true, get: function () { return pg_2.types; } });
__exportStar(require("./core"), exports);
exports.default = pg_1.default;
